waiter/src/waiter/handler.clj

```diff
@@ -1174,14 +1174,52 @@
   [request]
   (utils/exception->response (ex-info (utils/message :not-found) {:log-level :info :status http-404-not-found}) request))
 
+(defn drain-handler
+  "Handles Get /drain requests by providing the current drain state.
+  Handles Post /drain requests and sets the drain-until for the router based on the provided 'drain-timeout-ms' query
+  param which defaults to 120 seconds. This will cause the /status health check to fail and remove the router in general
+  load balancers. The request may have the 'crash' query parameter, which when true, will cause the router to exit after
+  the provided timeout. You can cancel the 'crash' request by sending a subsequent request setting it back to false before
+  the 'drain-timeout-ms' is reached."
+  [clock drain-atom admin-user?-fn crash-fn drain-mode?-fn {:keys [request-method] :as request}]
+  (try
+    (case request-method
+      :get (utils/clj->json-response {:result @drain-atom
+                                      :drain-mode? (drain-mode?-fn)})
+      :post (let [{:strs [drain-timeout-ms] :or {drain-timeout-ms "120000"} :as request-params} (-> request ru/query-params-request :query-params)
+                  auth-user (get request :authorization/user)
+                  drain-timeout-ms (utils/parse-int drain-timeout-ms)
+                  crash-process? (utils/request-flag request-params "crash")
+                  now (clock)
+                  drain-until (t/plus now (t/millis drain-timeout-ms))]
+              (when (not (admin-user?-fn auth-user))
+                (throw (ex-info "Must be an admin to use this endpoint."
+                                {:auth-user auth-user :log-level :info :status http-401-unauthorized})))
```

Comment:
Should use `403 Forbidden` (indicates that authorization failed), not `401` (indicates that authentication failed).